### ISSUE TITLE
Java 8 in CI: the end is nigh

### ISSUE THREAD

**Issue by armanbilge** [typelevel/sbt-typelevel#215]
Maybe not tomorrow or even this year, but this stuckfast position will gradually grow more and more uncomfortable. E.g. our sbt-plugins are itching to leave us behind:
- https://github.com/sbt/sbt-git/issues/213

FWIW we can probably get along just fine without JDK 8 in CI. But perhaps I haven't heard the horror stories 😆 
1. The `tlJdkRelease := Some(8)` setting enables actives all the scalac and javac flags to safely build/release from a newer JDK.
2. fs2 only runs CI on JDK 17 and besides the occasional bump is really a success story. Note that the `tlJdkRelease` setting would actually prevent those `ByteBuffer`-oopsies. In fact fs2 _can't_ use the `tlJdkRelease` setting because it blocks access to JDK 9+ APIs!

**Comment by rossabaker** [comment]
Java 8 Active Support ended on March 31, which is historically the date I become less of an curmudgeon about maintaining old things.

**Comment by armanbilge** [comment]
Politely pinging @vasilmkd, do you have any opinions on this? Things are still a bit new with the `tlJdkRelease` flag on Scala 3 but actually so far it seems to be doing _better_ than vanilla JDK 8 🤣

**Comment by vasilmkd** [comment]
I wouldn't completely drop it in Cats Effect (and we would set that as a special case), but as a default, I think we should be nudging users towards the future and make JDK 17 the default.

**Comment by armanbilge** [comment]
Yup, this is definitely where things are headed. /cc @valencik
- https://github.com/scalameta/metals/pull/3788

In v0.5.0 we'll set `tlJdkRelease := 8`, switch the default CI JDK to 17, and upgrade the nix flake in the g8 template as well.

**Comment by armanbilge** [comment]
@bpholt thanks, that's helpful input. Projects such as Cats, Cats Effect, and Http4s will almost certainly want to keep the Java 8 builds as well.

So I guess we have to be careful to make sure sbt-typelevel still works on JDK 8, even if we switch the default CI to JDK 17 and set the release flag to Java 8 as proposed here.

We can keep thinking about this as we approach sbt-typelevel v0.5.x.

**Comment by armanbilge** [comment]
Another interesting idea:

So sbt has a setting for running scalac and forked tests on a different JDK than what sbt itself is running on (h/t @keynmol).
https://github.com/sbt/sbt/blob/5532af17c795e0e9981b542810f449035c25172e/main/src/main/scala/sbt/Keys.scala#L290

```scala
val javaHome = settingKey[Option[File]]("Selects the Java installation used for compiling and forking.  If None, uses the Java installation running the build.").withRank(ASetting)
```

If we can figure out how to configure this, then it would mean that our _tooling_ can upgrade to newer JDKs (such as sbt plugins as mentioned in https://github.com/typelevel/sbt-typelevel/issues/215#issue-1173136548) while we still keep compiling and testing our libraries on the ancient JDK of our choice.

**Comment by bpholt** [comment]
Another thing I just encountered: I don't think setting `-release` helps catch errors where a dependency is inadvertently updated to a version that requires a higher JDK version than the one being targeted by your project: e.g. [logback 1.3 targets Java 8 but logback 1.4 targets Java 11](https://logback.qos.ch/news.html).

I just had a CI run on Java 8 catch the inadvertent update (because tests failed due to the `java.lang.UnsupportedClassVersionError`). When I ran the tests with Java 11 locally, no error is raised.

If tests can continue to run on the old JDK like @armanbilge [suggested](https://github.com/typelevel/sbt-typelevel/issues/215#issuecomment-1303632754) that feels like solid middle ground. Or maybe there are other ways to catch this earlier?

**Comment by armanbilge** [comment]
> I don't think setting `-release` helps catch errors where a dependency is inadvertently updated to a version that requires a higher JDK version than the one being targeted by your project

Yes that's right. `-release` is a compiler flag, so it only affects how your code compiles. It can prevent you from calling certain JDK APIs. But it won't check your dependencies, and it can't stop you from calling APIs in dependencies that need newer JDKs.

---

I'm currently leaning towards the `javaHome` idea. I like that it means we can upgrade our build tooling relieve some of the upstream pressure while still running tests on the JDK of our choice.

It will however take some acrobatics to make this work with the existing `download-java` and `setup-java` actions. Basically we need a way to install at least two JDKs: one for sbt, and one for forked tests.

**Comment by armanbilge** [comment]
I still like the `javaHome` idea but it's not going to land in 0.5.0.

**Comment by keynmol** [comment]
With regards to java home, you can theoretically use something like this: https://github.com/sourcegraph/scip-java/blob/main/project/JavaToolchainPlugin.scala

I should warn you though that when using SBT BSP with Metals, some java compilation state gets corrupted and a lot of weird artifacts appear.

The concept replicates Gradle's toolchains, and apart from this weirdness works reasonably well.

**Comment by armanbilge** [comment]
mdoc just dropped Java 8.
- https://github.com/scalameta/mdoc/releases/tag/v2.5.0

**Comment by armanbilge** [comment]
I think we no longer support Java 8 for users on the new Sonatype Central host (not the legacy or s01 hosts).

- https://github.com/xerial/sbt-sonatype/issues/548

**Comment by djx314** [comment]
@armanbilge I think `settingKey` about `javahome` must bring to sbt itself. Java8 is the problem of project. It‘s not the problem with sbt-plugin.

**Comment by djx314** [comment]
If bring jdk configuration to sbt plugins. I have plan to catch all the `OpenJDK Download Links`. And codegen them to typesafe scala code. Then publish new version to maven use ci when `JDK Download LInks` changed.

Here's the sample code:
``` sbt
Then config the JDK version in build.sbt
``` sbt
import com.github.sbt.jdk.collection.JDKCollection
javahome := {
  // step 1
  val JdkVersion = JDKCollection.`v1.8`.`Liberica`.lastest
  // ↑ 2 style ↓
  val JdkVersion = JDKCollection.`v1.8`.`Liberica`.`v8u432`

  // step 2
  val jdkAction = (CurrentOS / JdkVersion).value
  val jdksDirectory = jdkAction.getJDKCollectionDirectory // not use in this code snippet // rootProjectDirectory/target/.jdks/
  val currentJDKHome = jdkAction.getHome // rootProjectDirectory/target/.jdks/Liberica/version_1_8/8u432/

  // step 3
  jdkAction.downladAndUnzipIfNotExists(currentJDKHome)

  // step 4
  Some(currentJDKHome)
}
```
```
